corgidrp/calibrate_nonlin.py

```diff
@@ -0,0 +1,960 @@
+    if not isinstance(nonlin_params['colroi1'], (float, int)):
+        raise TypeError('colroi1 is not a number')
+    if not isinstance(nonlin_params['colroi2'], (float, int)):
+        raise TypeError('colroi2 is not a number')
+    if not isinstance(nonlin_params['rowback11'], (float, int)):
+        raise TypeError('rowback11 is not a number')
+    if not isinstance(nonlin_params['rowback12'], (float, int)):
+        raise TypeError('rowback12 is not a number')
+    if not isinstance(nonlin_params['rowback21'], (float, int)):
+        raise TypeError('rowback21 is not a number')
+    if not isinstance(nonlin_params['rowback22'], (float, int)):
+        raise TypeError('rowback22 is not a number')
+    if not isinstance(nonlin_params['colback11'], (float, int)):
+        raise TypeError('colback11 is not a number')
+    if not isinstance(nonlin_params['colback12'], (float, int)):
+        raise TypeError('colback12 is not a number')
+    if not isinstance(nonlin_params['colback21'], (float, int)):
+        raise TypeError('colback21 is not a number')
+    if not isinstance(nonlin_params['colback22'], (float, int)):
+        raise TypeError('colback22 is not a number')
+    if not isinstance(nonlin_params['min_exp_time'], (float, int)):
+        raise TypeError('min_exp_time is not a number')
+    if not isinstance(nonlin_params['num_bins'], (float, int)):
+        raise TypeError('num_bins is not a number')
+    if not isinstance(nonlin_params['min_bin'], (float, int)):
+        raise TypeError('min_bin is not a number')
+    if not isinstance(nonlin_params['min_mask_factor'], (float, int)):
+        raise TypeError('min_mask_factor is not a number')
+    
+
+class CalNonlinException(Exception):
+    """Exception class for calibrate_nonlin."""
+
+def frameProc(frame, offset_colroi, emgain):
+    """
+    simple row-bias subtraction using prescan region
+    frame of an L1 SCI-size frame, offset_colroi is the
+    column range in the prescan region to use to calculate
+    the median for each row. EM gain is the actual emgain used
+    to collect the frame.
+
+    Args:
+      frame (np.array): L1 frame
+      offset_colroi (int): column range
+      emgain (int): EM gain value
+
+    Returns:
+      np.array: bias subtracted frame
+    """
+
+    frame = np.float64(frame)
+    row_meds = np.median(frame[:,offset_colroi], axis=1)
+    row_meds = row_meds[:, np.newaxis]
+    frame -= row_meds
+    frame = frame/emgain
+    return frame
+
+def calibrate_nonlin(dataset_nl, actual_gain_arr, actual_gain_mean_frame,
+                     n_cal=20, n_mean=30, norm_val = 2500, min_write = 800.0, max_write = 10000.0,
+                     lowess_frac = 0.1, rms_low_limit = 0.004, rms_upp_limit = 0.2,
+                     pfit_upp_cutoff1 = -2, pfit_upp_cutoff2 = -3,
+                     pfit_low_cutoff1 = 2, pfit_low_cutoff2 = 1,
+                     make_plot=True, plot_outdir='figures', show_plot=False,
+                     verbose=False):
```

Comment:
Same comment as for kgain: the gains we can get from the headers and don't need to be input arguments here. For now `CMDGAIN` is appropriate since we haven't implemented actual gain.

Reply:
Done